src/touch.rs

```diff
@@ -0,0 +1,100 @@
+use esp_idf_sys::{
+    esp, touch_fsm_mode_t, touch_fsm_mode_t_TOUCH_FSM_MODE_MAX, touch_fsm_mode_t_TOUCH_FSM_MODE_SW,
+    touch_fsm_mode_t_TOUCH_FSM_MODE_TIMER, touch_pad_config, touch_pad_fsm_start, touch_pad_init,
+    touch_pad_read_raw_data, touch_pad_set_fsm_mode, touch_pad_t, touch_pad_t_TOUCH_PAD_NUM0,
+    touch_pad_t_TOUCH_PAD_NUM1, touch_pad_t_TOUCH_PAD_NUM10, touch_pad_t_TOUCH_PAD_NUM11,
+    touch_pad_t_TOUCH_PAD_NUM12, touch_pad_t_TOUCH_PAD_NUM13, touch_pad_t_TOUCH_PAD_NUM14,
+    touch_pad_t_TOUCH_PAD_NUM2, touch_pad_t_TOUCH_PAD_NUM3, touch_pad_t_TOUCH_PAD_NUM4,
+    touch_pad_t_TOUCH_PAD_NUM5, touch_pad_t_TOUCH_PAD_NUM6, touch_pad_t_TOUCH_PAD_NUM7,
+    touch_pad_t_TOUCH_PAD_NUM8, touch_pad_t_TOUCH_PAD_NUM9, EspError,
+};
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub enum FsmMode {
+    Timer,
+    SW,
+    Max,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl From<FsmMode> for touch_fsm_mode_t {
+    fn from(value: FsmMode) -> Self {
+        match value {
+            FsmMode::Timer => touch_fsm_mode_t_TOUCH_FSM_MODE_TIMER,
+            FsmMode::SW => touch_fsm_mode_t_TOUCH_FSM_MODE_SW,
+            FsmMode::Max => touch_fsm_mode_t_TOUCH_FSM_MODE_MAX,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub enum TouchPad {
+    Pad0,
+    Pad1,
+    Pad2,
+    Pad3,
+    Pad4,
+    Pad5,
+    Pad6,
+    Pad7,
+    Pad8,
+    Pad9,
+    Pad10,
+    Pad11,
+    Pad12,
+    Pad13,
+    Pad14,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl From<TouchPad> for touch_pad_t {
+    fn from(value: TouchPad) -> Self {
+        match value {
+            TouchPad::Pad0 => touch_pad_t_TOUCH_PAD_NUM0,
+            TouchPad::Pad1 => touch_pad_t_TOUCH_PAD_NUM1,
+            TouchPad::Pad2 => touch_pad_t_TOUCH_PAD_NUM2,
+            TouchPad::Pad3 => touch_pad_t_TOUCH_PAD_NUM3,
+            TouchPad::Pad4 => touch_pad_t_TOUCH_PAD_NUM4,
+            TouchPad::Pad5 => touch_pad_t_TOUCH_PAD_NUM5,
+            TouchPad::Pad6 => touch_pad_t_TOUCH_PAD_NUM6,
+            TouchPad::Pad7 => touch_pad_t_TOUCH_PAD_NUM7,
+            TouchPad::Pad8 => touch_pad_t_TOUCH_PAD_NUM8,
+            TouchPad::Pad9 => touch_pad_t_TOUCH_PAD_NUM9,
+            TouchPad::Pad10 => touch_pad_t_TOUCH_PAD_NUM10,
+            TouchPad::Pad11 => touch_pad_t_TOUCH_PAD_NUM11,
+            TouchPad::Pad12 => touch_pad_t_TOUCH_PAD_NUM12,
+            TouchPad::Pad13 => touch_pad_t_TOUCH_PAD_NUM13,
+            TouchPad::Pad14 => touch_pad_t_TOUCH_PAD_NUM14,
+        }
+    }
+}
+
+pub struct TouchConfig {
+    fsm_mode: FsmMode,
+    configured_pads: Vec<TouchPad>,
```

Comment:
This `Vec` seems to be causing build issues, not sure how best to resolve?

Reply:
why do you start collecting all the pads in one struct anyway. In the end from a design perspective it will be problematic, for example if you want to set or read a pin, or if you want to create an ISR/callback for it, you somehow need to be able to address that pin. So a general approach would be here similar to the PinDriver - where each PinDriver is one instance tight to one gpio pin. 

If you have problems that some part of the api should only be called once like some of the init code we maybe need to split them of ( need to find that out)  - in that case we would have something like a TouchDriver that would do the innit and config stuff throughout all pads - and them something like a TouchPadDriver for each individual pad to get to the individual get / set methods etc. Otherwise if thats not a problem just go with one struct.

Reply:
So `init` is universal so I think we will need both, how do you envisage `TouchDriver` and `TouchPadDriver` interacting? Create the latter from a method on the former? The FSM stuff is also universal and I'm not sure if that needs to be called after conflagration of individual pads?

Reply:
one thing we care about is cleaning up behind ourself automatically. So if a PadDriver goes out of scope it needs to be dropped correctly. We could make it that a TouchPadDriver gets a &TouchDriver and so it gets the lifetime of it intrinsically. Also you need to make sure to implement Drop correctly on both, so if TouchDriver gets out of scope the driver gets de-init correctly in reverse order.

Reply:
Still not totally clear on what's needed here, is there analogous code in some other driver I could look at for inspiration?